### PR TITLE
Hot module replacement

### PR DESCRIPTION
**Pull request by phlmn**

### CODE REVIEW

File: src/modules.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+export function setCurrentModule(mod: string | null) {
+  currentModule = mod;
+}
+
+type CleanupFn = () => void | Promise<void>;
+let cleanupFns: Array<{ fn: CleanupFn, module: string }> = [];
+
+export function registerModuleCleanup(fn: () => void | Promise<void>) {
+  if (!currentModule) {
+    // TODO: print warning? or collect cleanup that do not belong to a module?
```
Comment (anuejn) [inline src/modules.ts:22]: this should be implemented? at least a warning sounds like a good idea

File: src/index.ts [hunk 2]
```diff
@@ -2,68 +2,143 @@ export * from './state';
...
+    delete require.cache[require.resolve(file)];
   }
 
   for (const file of siteFiles) {
     try {
-      const modulePath = path.join(process.cwd(), file);
-      console.debug(`Loading module '${moduleName(file)}'`);
-      setCurrentModule(moduleName(file));
-      require(modulePath);
+      // const modulePath = path.join(process.cwd(), file);
```
Comment (anuejn) [inline src/index.ts]: :recycle:

File: src/index.ts [hunk 3]
```diff
@@ -84,8 +159,11 @@ export async function run(siteRoot: string = `${process.cwd()}/site`) {
 
   // const files = await glob(`${siteRoot}/**/*.ts*`);
```
Comment (anuejn) [inline src/index.ts]: :recycle:

**Review by anuejn — CHANGES_REQUESTED** [review]

File: src/modules.ts [hunk 4]
```diff
@@ -95,3 +120,21 @@ export function moduleName(fileName: string, rootFolder: string) {
...
+  const base = path.join(rootFolder, module);
+
+  for (let ext of EXTENSIONS) {
+    const filePath = base + '.' + ext;
+    try {
+      await stat(base + '.' + ext);
+      return filePath;
+    }
+    catch {
+      // no-op
```
[the remark this answers is not in the document]
Comment (phlmn) [reply src/modules.ts:136]: Yes, we skip files that do not exit and if non could be found an error is thrown on line 140:

```js
throw Error('Source file not found');
```

Sadly `exits(…)` is deprecated...

File: src/index.ts [hunk 5]
```diff
@@ -2,68 +2,143 @@ export * from './state';
...
+  const mainFun = require(path.join(rootFolder, 'index.js')).default;
+  await mainFun();
   setCurrentModule(null);
   console.log();
 
-  // HACK: wait until the application is initialized before loading site modules
-  await new Promise((resolve) => {
-    setTimeout(resolve, 100)
-  });
+  // // HACK: wait until the application is initialized before loading site modules
```
[the remark this answers is not in the document]
Comment (phlmn) [reply src/index.ts]: this is outdated

File: src/index.ts (hunk 2 above)
Comment (phlmn) [reply src/index.ts]: this is outdated

File: src/index.ts (hunk 3 above)
Comment (phlmn) [reply src/index.ts]: this is outdated

File: example/index.ts [hunk 6]
```diff
@@ -6,7 +6,7 @@ async function main() {
 
   // this function collects all typescript files from the `site/` folder and runs them.
   // this might feel a bit 'magic' at first but is rather handy when adding / removing code.
-  run();
+  // run();
```
[the remark this answers is not in the document]
Comment (phlmn) [reply example/index.ts]: this is outdated

File: example/package.json [hunk 7]
```diff
@@ -3,14 +3,16 @@
   "version": "0.1.0",
   "description": "An example how to use HAAH",
   "scripts": {
-    "start": "ts-node-dev index.ts"
+    "start": "haah"
```
[the remark this answers is not in the document]
Comment (phlmn) [reply example/package.json]: done

File: src/modules.ts (hunk 1 above)
Comment (phlmn) [reply src/modules.ts]: done

File: src/index.ts [hunk 8]
```diff
@@ -44,14 +45,16 @@ async function onBuild(
   );
   console.log();
 
-  cleanupModule('index');
+  if (initial) {
```
[the remark this answers is not in the document]
Comment (phlmn) [reply src/index.ts]: Good point! I've just created #33 for this :)

**Pull request merged**